content/blog/nielx/2022-12-16_http_network_services_preview_in_r1_beta_4.md

```diff
@@ -0,0 +1,109 @@
+| **Range-based Iteration** | [`BHttpFields`](https://git.haiku-os.org/haiku/tree/src/kits/network/libnetservices2/HttpFields.cpp?h=r1beta4#n477) | The [range-based `for` loop](https://learn.microsoft.com/en-us/cpp/cpp/range-based-for-statement-cpp?view=msvc-170) syntax is supported by the `BHttpFields` class. This allows you to loop through all the header fields in the object using this construct. Internally, it is used in most cases where there are loops through containers. |
+| **String Views** | [`BHttpMethod::BHttpMethod()`](https://git.haiku-os.org/haiku/tree/src/kits/network/libnetservices2/HttpRequest.cpp?h=r1beta4#n70) | The [`std::string_view`](https://en.cppreference.com/w/cpp/string/basic_string_view) object is a very simple concept: it is a pointer to a string, with a length embedded. This makes the use of borrowing (parts of) strings a lot simpler and safer, and has a lot of potential in the rest of the API. It is probably a good opportunity to investigate if this can be supported in the Haiku API more (or if it is needed to create a custom variant). |
+| **Optional** | [`BHttpRequest::SetRequestBody()`](https://git.haiku-os.org/haiku/tree/src/kits/network/libnetservices2/HttpRequest.cpp?h=r1beta4#n349) | The [`std::optional`](https://en.cppreference.com/w/cpp/utility/optional) utility allows you to express the situation where something either has a value, or it does not. It is useful both for return values, data structure members, and function arguments. In the example, it is used to handle the situation where sometimes you know the size of the request body, and therefore you can submit an `off_t` size argument, but sometimes you do not, and then you can express it by passing `std::nullopt`. Previously this would have required overloaded methods, or by overloading the meaning of the the size argument so that `0` equates to unknown. |
+| **Smart Pointers** | [`BHttpRequest::SetRequestBody()`](https://git.haiku-os.org/haiku/tree/src/kits/network/libnetservices2/HttpRequest.cpp?h=r1beta4#n349) | C++'s `std::unique_ptr` and `std::shared_ptr` are a great help to reducing data leaks by making it explicit who owns an object on the heap. They are amazing when [used correctly](https://herbsutter.com/2013/05/29/gotw-89-solution-smart-pointers/)! In the example, it is used to move ownership of a data stream (like a file) to the request. If the request then goes out of scope, it automatically cleans up the data stream object. More importantly, smart pointers and their allocation mechanisms (like [`std::make_unique()`](https://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique) ) are used throughout the internal implementation. In fact, you will not be able to find a single `new` or `delete` in the libnetservices2 source code! |
```

Comment:
We already have BReference (similar to shared_ptr) and BAutoDeleter (similar to unique_ptr). These at least deserve a mention and maybe also an explanation of the differences

Reply:
Will do.